packages/xo-server/src/xo-mixins/xen-servers.js

```diff
@@ -352,14 +355,40 @@ export default class {
 
     xapi.xo.install()
 
-    await xapi.connect().then(
-      () => this.updateXenServer(id, { error: null }),
-      error => {
-        this.updateXenServer(id, { error: serializeError(error) })
-
-        throw error
+    const onRedirect = ({ protocol, hostname, port }) => {
+      // FIXME: race condition with the master's connection at the xo-server restart
+      const isPoolMasterConnected = some(
+        this._xapis,
+        xapi =>
+          xapi.pool !== null &&
+          hostname === xapi.getObject(xapi.pool.master).address
+      )
+      if (isPoolMasterConnected) {
+        throw new Error("The server's master is already connected.")
       }
-    )
+
+      let host = ''
+      protocol !== undefined && (host += `${protocol}//`)
+      host += hostname
+      port !== undefined && (host += `:${port}`)
+      this.updateXenServer(id, {
+        allowConnected: true,
+        host,
+      })::ignoreErrors()
+    }
+
+    xapi.on('redirect', onRedirect)
+    await xapi
+      .connect()
+      .then(
+        () => this.updateXenServer(id, { error: null }),
+        error => {
+          this.updateXenServer(id, { error: serializeError(error) })
```

Comment:
[Question not related to this issue/PR]: should we delete the server entry in the `this._xapi` collection in this case ?

Reply:
Absolutely.